components/Ripple/src/MDCStatefulRippleView.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCRippleView.h"
+
+/**
+ Provides the current state of the ripple. The ripple is either in its normal state, or in the
+ selected state where the ripple remains spread on the view.
+
+ - MDCRippleStateNormal: The ripple isn't currently presented.
+ - MDCRippleStateHighlighted: The ripple is activated and shown.
+ - MDCRippleStateSelected: The ripple is in the selected state.
+ - MDCRippleStateDragged: The ripple is in the dragged state.
+ */
+typedef NS_OPTIONS(NSInteger, MDCRippleState) {
+  MDCRippleStateNormal = 0,
+  MDCRippleStateHighlighted = 1 << 0,
+  MDCRippleStateSelected = 1 << 1,
+  MDCRippleStateDragged = 1 << 2,
+};
+
+/**
+ This class subclasses MDCRippleView which provides the Ripple functionality, and adds
+ to it support for states. By providing states, you can set the overlay color for each state, and
+ change the state of the RippleView by using its highlighted/selected/dragged API.
+
+ This class, as opposed to MDCRippleView is less passive in its appraoch and listens to touches to
+ be able to record the last touch and initiate the Ripple from there, as opposed to receiving
+ the touch as input. It also is able to identify when a tap goes off/on the view, and fades in/out
+ the ripple accordingly.
+
+ Lastly, the implementation for states follows Material Guidelines closely and is built to work
+ well with UICollectionViewCells, UITableViewCells, and UIControls. Therefore MDCStatefulRippleView
+ should be used when one wants to leverage the Material state system and should ideally be
+ configured/set alongside the UIKit APIs (i.e. UIControlState or cell's setSelected/setHighlighted).
+ */
+@interface MDCStatefulRippleView : MDCRippleView
```

Comment:
Consider marking this class as `final` until we have a clear need to open the API up to subclassing.

Reply:
Done.